File: lib/screensaver.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'package:flutter/foundation.dart';

import 'player.dart';

// Manages xscreensaver inhibition based on video playback state
// Checks both mpv (external player) and browser (via CDP) for active video
class ScreensaverInhibitor {
  static ScreensaverInhibitor? _instance;

  Timer? _timer;
  int? _debugPort;
  String? _browserName;
  bool? _isNativeLinux;
  int _checkIntervalSeconds = 60; // Default, will be updated from xscreensaver config

  static ScreensaverInhibitor getInstance() {
    _instance ??= ScreensaverInhibitor();
    return _instance!;
  }

  // Set the browser info for video detection
  void setBrowser(String? browserName, int? port) {
    _browserName = browserName;
    _debugPort = port;
    // Check immediately when browser is set
    if (browserName != null && port != null) {
      _checkAndUpdate();
    }
  }

  // Start the periodic check (call this when app starts)
  Future<void> start() async {
    if (_timer != null) return;

    // Deactivate xscreensaver on startup in case it's currently running
    await _inhibit();

    // Read xscreensaver timeout and set check interval to half of it
    await _readXscreensaverTimeout();

    _timer = Timer.periodic(Duration(seconds: _checkIntervalSeconds), (_) {
      _checkAndUpdate();
    });

    // Also check immediately
    _checkAndUpdate();
  }

  // Stop the periodic check (call this when app exits)
  void stop() {
    _timer?.cancel();
    _timer = null;
  }

  // Read timeout from ~/.xscreensaver and set interval to half of it
  Future<void> _readXscreensaverTimeout() async {
    try {
      final home = Platform.environment['HOME'];
      if (home == null) return;

      final configFile = File('$home/.xscreensaver');
      if (!await configFile.exists()) return;

      final content = await configFile.readAsString();
      // Look for line like "timeout: 0:05:00" (hours:minutes:seconds)
      final match = RegExp(r'timeout:\s*(\d+):(\d+):(\d+)').firstMatch(content);
      if (match != null) {
        final hours = int.parse(match.group(1)!);
        final minutes = int.parse(match.group(2)!);
        final seconds = int.parse(match.group(3)!);
        final totalSeconds = hours * 3600 + minutes * 60 + seconds;

        // Set check interval to half the timeout, minimum 30 seconds
        _checkIntervalSeconds = (totalSeconds ~/ 2).clamp(30, 300);
        debugPrint('ScreensaverInhibitor: xscreensaver timeout is $totalSeconds seconds, checking every $_checkIntervalSeconds seconds');
      }
    } catch (e) {
      debugPrint('ScreensaverInhibitor: Failed to read xscreensaver config: $e');
    }
  }

  Future<void> _checkAndUpdate() async {
    // Only run on native Linux (not WSL, not Windows)
    if (!await _isOnNativeLinux()) return;

    final isPlaying = await _isVideoPlaying();

    if (isPlaying) {
      await _inhibit();
    }
  }

  Future<bool> _isOnNativeLinux() async {
    if (_isNativeLinux != null) return _isNativeLinux!;

    if (Platform.isWindows) {
      _isNativeLinux = false;
      return false;
    }

    // Check for WSL
    try {
      final versionFile = File('/proc/version');
      if (await versionFile.exists()) {
        final content = await versionFile.readAsString();
        if (content.toLowerCase().contains('microsoft')) {
          _isNativeLinux = false;
          return false;
        }
      }
    } catch (_) {}

    _isNativeLinux = true;
    return true;
  }

  Future<bool> _isVideoPlaying() async {
    // Check mpv first
    final player = ExternalPlayer.getInstance();
    if (player.isPlaying && !player.paused) {
      return true;
    }

    // Check browser via remote debugging
    if (_debugPort != null && _browserName != null) {
      try {
        final playing = await _checkBrowserVideoPlaying();
        if (playing) return true;
      } catch (_) {
        // Browser check failed, ignore
      }
    }

    return false;
  }

  Future<bool> _checkBrowserVideoPlaying() async {
    if (_debugPort == null) {
      debugPrint('ScreensaverInhibitor: No debug port set');
      return false;
    }

    debugPrint('ScreensaverInhibitor: Checking browser "$_browserName" on port $_debugPort');

    if (_browserName == 'Firefox') {
      final result = await _checkFirefoxVideoPlaying();
      debugPrint('ScreensaverInhibitor: Firefox fullscreen video playing: $result');
      return result;
    } else {
      final result = await _checkChromeVideoPlaying();
      debugPrint('ScreensaverInhibitor: Chrome fullscreen video playing: $result');
      return result;
    }
  }

  // Chrome DevTools Protocol
  Future<bool> _checkChromeVideoPlaying() async {
    try {
      final client = HttpClient();
      final request = await client.getUrl(
        Uri.parse('http://localhost:$_debugPort/json/list'),
      ).timeout(const Duration(seconds: 2));
      final response = await request.close();

      if (response.statusCode != 200) {
        client.close();
        return false;
      }

      final body = await response.transform(utf8.decoder).join();
      final targets = jsonDecode(body) as List<dynamic>;
      client.close();

      // Find a page target to query
      for (final target in targets) {
        if (target['type'] == 'page') {
          final wsUrl = target['webSocketDebuggerUrl'] as String?;
          if (wsUrl != null) {
            final playing = await _checkChromePageForVideo(wsUrl);
            if (playing) return true;
          }
        }
      }
    } catch (e) {
      debugPrint('ScreensaverInhibitor: Chrome CDP check failed: $e');
    }

    return false;
  }

  // Firefox WebDriver BiDi protocol
  Future<bool> _checkFirefoxVideoPlaying() async {
    WebSocket? ws;
    try {
      // Connect to WebDriver BiDi WebSocket on /session path
      ws = await WebSocket.connect('ws://127.0.0.1:$_debugPort/session')
          .timeout(const Duration(seconds: 2));

      final completer = Completer<bool>();
      String? contextId;
      bool? videoPlayingResult;

      ws.listen((data) {
        try {
          final response = jsonDecode(data as String);
          debugPrint('ScreensaverInhibitor: Firefox BiDi response: $response');

          // Handle session.new response
          if (response['id'] == 1 && response['result'] != null) {
            debugPrint('ScreensaverInhibitor: Session created, getting tree');
            // Session created, now get the browsing contexts
            final getTreeCommand = jsonEncode({
              'id': 2,
              'method': 'browsingContext.getTree',
              'params': {},
            });
            ws?.add(getTreeCommand);
          }
          // Handle browsingContext.getTree response
          else if (response['id'] == 2 && response['result'] != null) {
            final contexts = response['result']['contexts'] as List<dynamic>?;
            if (contexts != null && contexts.isNotEmpty) {
              // Get the first top-level context
              contextId = contexts[0]['context'] as String?;
              debugPrint('ScreensaverInhibitor: Got context: $contextId');

              if (contextId != null) {
                // Now evaluate script to check for playing fullscreen videos
                final evalCommand = jsonEncode({
                  'id': 3,
                  'method': 'script.evaluate',
                  'params': {
                    'expression': '''
                      (function() {
                        // Only inhibit if there's a playing video AND we're in fullscreen
                        if (!document.fullscreenElement) {
                          return false;
                        }
                        const videos = document.querySelectorAll('video');
                        for (const v of videos) {
                          if (!v.paused && !v.ended && v.readyState > 2) {
                            return true;
                          }
                        }
                        return false;
                      })()
                    ''',
                    'target': {'context': contextId},
                    'awaitPromise': false,
                  },
                });
                ws?.add(evalCommand);
              } else {
                if (!completer.isCompleted) completer.complete(false);
              }
            } else {
              if (!completer.isCompleted) completer.complete(false);
            }
          }
          // Handle script.evaluate response
          else if (response['id'] == 3 && response['result'] != null) {
            final result = response['result']['result'];
            final value = result?['value'];
            debugPrint('ScreensaverInhibitor: Script result: $value');

            // Store the result, then end the session
            videoPlayingResult = value == true;

            // End the session before completing
            final endSessionCommand = jsonEncode({
              'id': 4,
              'method': 'session.end',
              'params': {},
            });
            ws?.add(endSessionCommand);
          }
          // Handle session.end response - now we can complete
          else if (response['id'] == 4) {
            debugPrint('ScreensaverInhibitor: Session ended');
            if (!completer.isCompleted) {
              completer.complete(videoPlayingResult ?? false);
            }
          }
          // Handle errors
          else if (response['error'] != null) {
            debugPrint('ScreensaverInhibitor: BiDi error: ${response['error']}');
            if (!completer.isCompleted) completer.complete(false);
          }
        } catch (e) {
          debugPrint('ScreensaverInhibitor: Firefox BiDi parse error: $e');
        }
      }, onError: (e) {
        debugPrint('ScreensaverInhibitor: Firefox BiDi socket error: $e');
        if (!completer.isCompleted) completer.complete(false);
      }, onDone: () {
        if (!completer.isCompleted) completer.complete(false);
      });

      // Start by creating a session
      final newSessionCommand = jsonEncode({
        'id': 1,
        'method': 'session.new',
        'params': {'capabilities': {}},
      });
      ws.add(newSessionCommand);

      // Timeout
      Future.delayed(const Duration(seconds: 3), () {
        if (!completer.isCompleted) completer.complete(false);
      });

      final result = await completer.future;
      await ws.close();
      return result;
    } catch (e) {
      debugPrint('ScreensaverInhibitor: Firefox BiDi check failed: $e');
      await ws?.close();
      return false;
    }
  }

  Future<bool> _checkChromePageForVideo(String wsUrl) async {
    try {
      final ws = await WebSocket.connect(wsUrl).timeout(const Duration(seconds: 2));

      // Query for playing fullscreen video
      final messageId = DateTime.now().millisecondsSinceEpoch;
      final command = jsonEncode({
        'id': messageId,
        'method': 'Runtime.evaluate',
        'params': {
          'expression': '''
            (function() {
              // Only inhibit if there's a playing video AND we're in fullscreen
              if (!document.fullscreenElement) {
                return false;
              }
              const videos = document.querySelectorAll('video');
              for (const v of videos) {
                if (!v.paused && !v.ended && v.readyState > 2) {
                  return true;
                }
              }
              return false;
            })()
          ''',
          'returnByValue': true,
        },
      });

      ws.add(command);

      // Wait for response
      final completer = Completer<bool>();
      Timer? timeout;

      timeout = Timer(const Duration(seconds: 2), () {
        if (!completer.isCompleted) {
          completer.complete(false);
        }
      });

      ws.listen((data) {
        try {
          final response = jsonDecode(data as String);
          if (response['id'] == messageId) {
            timeout?.cancel();
            final result = response['result']?['result']?['value'];
            if (!completer.isCompleted) {
              completer.complete(result == true);
            }
          }
        } catch (_) {}
      }, onError: (_) {
        timeout?.cancel();
        if (!completer.isCompleted) {
          completer.complete(false);
        }
      });

      final result = await completer.future;
      await ws.close();
      return result;
    } catch (e) {
      debugPrint('ScreensaverInhibitor: WebSocket check failed: $e');
      return false;
    }
  }

  Future<void> _inhibit() async {
    try {
      await Process.run('xscreensaver-command', ['-deactivate']);
    } catch (e) {
      debugPrint('ScreensaverInhibitor: Failed to inhibit: $e');
    }
  }
}

File: app/screensaver.go
package main

import (
	"bufio"
	"encoding/json"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"runtime"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/gorilla/websocket"
)

// ScreensaverInhibitor manages xscreensaver inhibition based on video playback
type ScreensaverInhibitor struct {
	mu            sync.Mutex
	player        *Player
	browserMgr    *BrowserManager
	ticker        *time.Ticker
	stopChan      chan struct{}
	checkInterval time.Duration
	isNativeLinux *bool
}

func NewScreensaverInhibitor(player *Player, browserMgr *BrowserManager) *ScreensaverInhibitor {
	return &ScreensaverInhibitor{
		player:        player,
		browserMgr:    browserMgr,
		checkInterval: 60 * time.Second, // Default, will be updated from config
	}
}

// Start begins periodic screensaver inhibition checks
func (s *ScreensaverInhibitor) Start() {
	s.mu.Lock()
	defer s.mu.Unlock()

	if s.ticker != nil {
		return // Already running
	}

	// Only run on native Linux
	if !s.isOnNativeLinux() {
		Log("ScreensaverInhibitor: Not on native Linux, disabled")
		return
	}

	// Read xscreensaver config for timeout
	s.readXscreensaverTimeout()

	// Deactivate on startup
	s.inhibit()

	s.stopChan = make(chan struct{})
	s.ticker = time.NewTicker(s.checkInterval)

	go func() {
		for {
			select {
			case <-s.ticker.C:
				s.checkAndUpdate()
			case <-s.stopChan:
				return
			}
		}
	}()

	Log("ScreensaverInhibitor: Started, checking every %v", s.checkInterval)
}

// Stop ends the periodic checks
func (s *ScreensaverInhibitor) Stop() {
	s.mu.Lock()
	defer s.mu.Unlock()

	if s.ticker != nil {
		s.ticker.Stop()
		close(s.stopChan)
		s.ticker = nil
		Log("ScreensaverInhibitor: Stopped")
	}
}

func (s *ScreensaverInhibitor) isOnNativeLinux() bool {
	if s.isNativeLinux != nil {
		return *s.isNativeLinux
	}

	result := false
	if runtime.GOOS == "linux" && !isWSL() {
		result = true
	}
	s.isNativeLinux = &result
	return result
}

// Read timeout from ~/.xscreensaver and set interval to half of it
func (s *ScreensaverInhibitor) readXscreensaverTimeout() {
	home, err := os.UserHomeDir()
	if err != nil {
		return
	}

	configPath := filepath.Join(home, ".xscreensaver")
	file, err := os.Open(configPath)
	if err != nil {
		return
	}
	defer file.Close()

	// Look for line like "timeout: 0:05:00" (hours:minutes:seconds)
	re := regexp.MustCompile(`timeout:\s*(\d+):(\d+):(\d+)`)
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		matches := re.FindStringSubmatch(line)
		if len(matches) == 4 {
			hours, _ := strconv.Atoi(matches[1])
			minutes, _ := strconv.Atoi(matches[2])
			seconds, _ := strconv.Atoi(matches[3])
			totalSeconds := hours*3600 + minutes*60 + seconds

			// Set check interval to half the timeout, minimum 30 seconds, max 300
			intervalSeconds := totalSeconds / 2
			if intervalSeconds < 30 {
				intervalSeconds = 30
			}
			if intervalSeconds > 300 {
				intervalSeconds = 300
			}
			s.checkInterval = time.Duration(intervalSeconds) * time.Second
			Log("ScreensaverInhibitor: xscreensaver timeout is %d seconds, checking every %d seconds", totalSeconds, intervalSeconds)
			return
		}
	}
}

func (s *ScreensaverInhibitor) checkAndUpdate() {
	if s.isVideoPlaying() {
		s.inhibit()
	}
}

func (s *ScreensaverInhibitor) isVideoPlaying() bool {
	// Check mpv player first
	if s.player != nil {
		status := s.player.GetStatus()
		if playing, ok := status["playing"].(bool); ok && playing {
			if paused, ok := status["paused"].(bool); ok && !paused {
				return true
			}
		}
	}

	// Check browser for fullscreen video via CDP
	if s.browserMgr != nil && s.browserMgr.IsRunning() {
		if s.checkBrowserVideoPlaying() {
			return true
		}
	}

	return false
}

func (s *ScreensaverInhibitor) checkBrowserVideoPlaying() bool {
	// Get CDP targets
	resp, err := http.Get("http://localhost:9222/json")
	if err != nil {
		return false
	}
	defer resp.Body.Close()

	var targets []struct {
		ID                   string `json:"id"`
		Type                 string `json:"type"`
		WebSocketDebuggerURL string `json:"webSocketDebuggerUrl"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&targets); err != nil {
		return false
	}

	// Check each page target for fullscreen video
	for _, t := range targets {
		if t.Type == "page" && t.WebSocketDebuggerURL != "" {
			if s.checkPageForVideo(t.WebSocketDebuggerURL) {
				return true
			}
		}
	}

	return false
}

func (s *ScreensaverInhibitor) checkPageForVideo(wsURL string) bool {
	conn, _, err := websocket.DefaultDialer.Dial(wsURL, nil)
	if err != nil {
		return false
	}
	defer conn.Close()

	conn.SetReadDeadline(time.Now().Add(2 * time.Second))
	conn.SetWriteDeadline(time.Now().Add(2 * time.Second))

	// Query for playing fullscreen video
	cmd := map[string]interface{}{
		"id":     1,
		"method": "Runtime.evaluate",
		"params": map[string]interface{}{
			"expression": `(function() {
				// Only inhibit if there's a playing video AND we're in fullscreen
				if (!document.fullscreenElement) {
					return false;
				}
				const videos = document.querySelectorAll('video');
				for (const v of videos) {
					if (!v.paused && !v.ended && v.readyState > 2) {
						return true;
					}
				}
				return false;
			})()`,
			"returnByValue": true,
		},
	}

	if err := conn.WriteJSON(cmd); err != nil {
		return false
	}

	var response struct {
		ID     int `json:"id"`
		Result struct {
			Result struct {
				Value interface{} `json:"value"`
			} `json:"result"`
		} `json:"result"`
	}

	if err := conn.ReadJSON(&response); err != nil {
		return false
	}

	if result, ok := response.Result.Result.Value.(bool); ok {
		return result
	}

	return false
}

func (s *ScreensaverInhibitor) inhibit() {
	cmd := exec.Command("xscreensaver-command", "-deactivate")
	if err := cmd.Run(); err != nil {
		// xscreensaver-command may not be installed or xscreensaver not running
		// This is fine, just means no screensaver to inhibit
		if !strings.Contains(err.Error(), "not found") {
			Log("ScreensaverInhibitor: Failed to deactivate: %v", err)
		}
	}
}
